File: adapters/python/src/postulate/load_spec.py
from __future__ import annotations

from pathlib import Path

import yaml
from pydantic import ValidationError

from postulate.models import PostulateSpec


class SpecLoadError(Exception):
    """Raised when a spec file cannot be loaded or validated."""


def _validate_parsed_spec(parsed: object, source: str) -> PostulateSpec:
    try:
        return PostulateSpec.model_validate(parsed)
    except ValidationError as err:
        issues = "\n".join(
            f"  - {'.'.join(str(part) for part in issue['loc']) or '<root>'}: {issue['msg']}"
            for issue in err.errors()
        )
        raise SpecLoadError(
            f"Spec {source} failed schema validation:\n{issues}"
        ) from err


def load_spec_from_content(content: str, source: str = "<string>") -> PostulateSpec:
    try:
        parsed = yaml.safe_load(content)
    except yaml.YAMLError as err:
        raise SpecLoadError(f"Invalid YAML in {source}: {err}") from err
    return _validate_parsed_spec(parsed, source)


def load_spec(spec_path: str | Path) -> PostulateSpec:
    path = Path(spec_path)
    abs_path = path.resolve()

    try:
        raw = abs_path.read_text(encoding="utf-8")
    except FileNotFoundError:
        raise SpecLoadError(f"Spec file not found: {abs_path}") from None
    except OSError as err:
        raise SpecLoadError(
            f"Could not read spec file {abs_path}: {err}"
        ) from err

    return load_spec_from_content(raw, str(spec_path))

File: src/loadSpec.ts
import fs from "node:fs";
import path from "node:path";
import YAML from "yaml";
import { z } from "zod";
import { PostulateSchema, type PostulateSpec } from "./spec.js";

export class SpecLoadError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "SpecLoadError";
  }
}

export function loadSpec(specPath: string): PostulateSpec {
  const abs = path.resolve(specPath);

  let raw: string;
  try {
    raw = fs.readFileSync(abs, "utf8");
  } catch (err) {
    const code = (err as NodeJS.ErrnoException).code;
    if (code === "ENOENT") {
      throw new SpecLoadError(`Spec file not found: ${abs}`);
    }
    throw new SpecLoadError(
      `Could not read spec file ${abs}: ${(err as Error).message}`
    );
  }

  return loadSpecFromContent(raw, specPath);
}

export function loadSpecFromContent(raw: string, source: string): PostulateSpec {
  let parsed: unknown;
  try {
    parsed = YAML.parse(raw);
  } catch (err) {
    throw new SpecLoadError(
      `Invalid YAML in ${source}: ${(err as Error).message}`
    );
  }

  try {
    return PostulateSchema.parse(parsed);
  } catch (err) {
    if (err instanceof z.ZodError) {
      const issues = err.issues
        .map((issue) => {
          const where = issue.path.length > 0 ? issue.path.join(".") : "<root>";
          return `  - ${where}: ${issue.message}`;
        })
        .join("\n");
      throw new SpecLoadError(
        `Spec ${source} failed schema validation:\n${issues}`
      );
    }
    throw err;
  }
}
